editors/code/src/commands/on_enter.ts

```diff
@@ -7,7 +7,7 @@ import { Cmd, Ctx } from '../ctx';
 async function handleKeypress(ctx: Ctx) {
     const editor = ctx.activeRustEditor;
     const client = ctx.client;
```

Comment:
```suggestion
    const client = ctx.client ?? return false;
```

Reply:
Heh ))((()))) I wish `return false` was an expression in JavaScript, but alas it isn't ...

Reply:
:man_facepalming:

Reply:
By the way, we could avoid this workaround by disposing command handlers on for the time of the server restart and reregistering them once the LSP is ready. @matklad did you consider that?

Reply:
I am not sure that this'll work in all contexts: commands in general are asynchronous, so we should be prepared that the client is disposed while some future is scheduled....

Not sure what's the proper way to deal with this though... Don't hold references to `client` across yield points seems like a bad solution.

Reply:
Hmm, right, and we do hold reference to the client once the execution of the handler started, but it can be disposed in the middle of the run of the handler anyway... I see in this situation we would need to use cancellation tokens otherwise even current setup has the flaw of holding a reference to shutdown LSP server.

Reply:
Or, we could use an async mutex, I already had experience with protecting async code with it. What do you think @matklad ?

Reply:
Or better async read-write lock...

Reply:
I just saw this comment 
https://github.com/rust-analyzer/rust-analyzer/blob/8d0f7da2f5f2ae1dc5711005f08fde0007da165b/editors/code/src/ctx.ts#L9-L13

and I see that you wanted to use async getter, which I never saw to be used in real code, however something like async read-write lock would be much more reliable.

Reply:
I think that *just* slapping a mutex on top of a thing won't help. I have not though yet what behavior do we want here in the first place.

Reply:
[Here is](https://github.com/mvisat/async-rwlock/blob/master/src/rwlock.ts) the minimal implementation which is distributed as a dependency-free npm package.

Reply:
I guess, the answer is: "we should do the same thing that VS Code does to ensure that `deactivate` is called only when there are no outstanding work scheduled by the extension". I don't how VS Code does this, or if it does this in the first place at all.

Reply:
Yeah, this is interesting know how vscode ensures that `deactivate()` is not called in the middle of the work scheduled by the extension. If I am not mistaken vscode creates a separate process for the extension, so it could just use `process.exit()` after the `deactivate()` and not wait for the scheduled work, but it is just my supposition.

However I think slapping the read-write lock on top of a thing would help. When the server is about to restart we acquire the `write()` operation, and all other handlers acquire `read()` operation so that before restarting the server we await that all handlers have executed the scheduled work, and also all new events will wait while the write lock is released.

Reply:
>However I think slapping the read-write lock on top of a thing would help indeed. When the server is about to restart we acquire the write() operation, and all other handlers acquire read() operation so that before restarting the server we await that all handlers have executed the scheduled work, and also all new events will wait while the write lock is not released.

This might cause live-lock where server does not start beacause new work is continuously scheduled. In general, I am usually not smart enough to convince myself that a mutex-based solution works. We really should check how VS Code handles this problem and follow the suite. This is unrelated to this PR though.

Reply:
Well, yes, I'll note that we need to checkout vscode solution. Though my gut feeling says we would not get a live lock because all acquire() operations would be queued and yes it may take some time before all pending read() handles are released but it won't be noticeable... Also see how restartServer() is not reentrant, I remember you talking about reentrancy at some CSC YT lecture ;p

Reply:
I wish we could avoid any kind of sync here, but se la vie

Reply:
@matklad, I did look into vscode `deactivate()` implementation. There is nothing fancy there...

https://github.com/microsoft/vscode/blob/1a81711a85e38ccf784110568ebf3784ab9094a5/src/vs/workbench/api/common/extHostExtensionService.ts#L253-L289

Reply:
Also, according to my assumptions, when run in a `browser` environment, the extension is started in a [dedicated worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers#Dedicated_workers) created [here](https://github.com/microsoft/vscode/blob/201b5a20571c4d7df816816297fb556ea56d58b0/src/vs/workbench/services/extensions/browser/webWorkerExtensionHostStarter.ts#L53-L106). The termination is pretty trivial, they provide 10 seconds before the abrupt termination of the thread [here](https://github.com/microsoft/vscode/blob/201b5a20571c4d7df816816297fb556ea56d58b0/src/vs/workbench/services/extensions/browser/webWorkerExtensionHostStarter.ts#L108-L119).

In `electron-browser` environment they do [fork a child process](https://github.com/microsoft/vscode/blob/1a81711a85e38ccf784110568ebf3784ab9094a5/src/vs/workbench/services/extensions/electron-browser/extensionHost.ts#L178). And they also [give 10 seconds before killing the process](https://github.com/microsoft/vscode/blob/201b5a20571c4d7df816816297fb556ea56d58b0/src/vs/workbench/services/extensions/electron-browser/extensionHost.ts#L518-L564).

Everything is protected by a boolean flag :D

Reply:
Much better would be to look at TypeScript `Restart TS Server` implementation.
First of, one not necessarily relevant, but interesting note. Not so long ago they introduced a new syntax-server separate from semmantic-server. You can see [code killing them both here](https://github.com/microsoft/vscode/blob/6ec6f9e3f4d17f0cef0480688a1f350b9141e567/extensions/typescript-language-features/src/tsServer/server.ts#L437-L440).

I didn't dive too deeply into their code, but it's obvious that they just [restart the server synchronously](https://github.com/microsoft/vscode/blob/6ec6f9e3f4d17f0cef0480688a1f350b9141e567/extensions/typescript-language-features/src/typescriptServiceClient.ts#L227-L235) via [kill()](https://github.com/microsoft/vscode/blob/6ec6f9e3f4d17f0cef0480688a1f350b9141e567/extensions/typescript-language-features/src/tsServer/spawner.ts#L242-L244) and [fork()](https://github.com/microsoft/vscode/blob/6ec6f9e3f4d17f0cef0480688a1f350b9141e567/extensions/typescript-language-features/src/tsServer/spawner.ts#L98) (beware that [`child_process.fork()` is not a POSIX fork at all](https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options)). And that `isRestarting` boolean is i no way a syncronization primitive, it's just used to be passed to [`serviceExited()` method, that sanity-checks the number of restarts](https://github.com/microsoft/vscode/blob/6ec6f9e3f4d17f0cef0480688a1f350b9141e567/extensions/typescript-language-features/src/typescriptServiceClient.ts#L534-L538).

I wish we could await promises from `vscode-languageclient` synchronously, but this is not possible...

Reply:
In omnisharp they have custom event stream that [is cleared on restart event](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/observers/OmnisharpChannelObserver.ts#L19). They use [rxjs](https://rxjs-dev.firebaseapp.com/) for [their EventStream implementation](https://github.com/OmniSharp/omnisharp-vscode/blob/4ca51555c2235f7c13bf835090f98322326c3ba6/src/EventStream.ts#L5), this is a very nice and popular library for reactive programming, but quite heavyweight (needs some additional time investment to learn)

Reply:
cc @matklad

Reply:
If TS does not care about clean shutdown of pending tasks, we shouldn't either.

Reply:
I wish I could say "We can do better!", and we can! But we don't want ;D